### ISSUE TITLE
CI: Run cpu tests in parallel processes?

### ISSUE THREAD

**Issue by peterbell10** [pytorch/pytorch#81110]
AFAICT CI is running on multi-core workers, but only launching a single process to run the tests. Since most individual test cases have low thread utilization, this means under-utilizing the available hardware.

For example, using [`pytest-xdist`](https://pypi.org/project/pytest-xdist/) I see a 4x speedup from running 8 parallel processes on a 32-core machine
```
$ time OMP_NUM_THREADS=32 pytest test_ops.py -k cpu
real	24m13.282s
user	369m12.584s
sys	11m12.221s

$ time OMP_NUM_THREADS=4 pytest test_ops.py -k cpu -n 8
real	5m22.487s
user	61m53.781s
sys	2m21.365s
```

There is some risk of additional flakiness. For example, two tests allocating large tensors might fail if run at the same time. This could be mitigated though, by marking problematic tests as not suitable for running in parallel.

**Comment by malfet** [comment]
@clee2000 have you tried limiting number of OMP threads before running your parallelisation experiments?

**Comment by clee2000** [comment]
yes, it didn't seem to affect the speedup when the number of processes was held constant, but I can try again with different numbers

@peterbell10 I'm actually currently looking into using pytest and pytest-xdist to run tests in parallel.  I have, as you pointed out, run into issues with memory on cuda tests, but cpu tests seem to be doing fine so far.  Do you know if pytest-xdist has something built in for not running specific tests in parallel or is this just something we'd have to do manually?  If you have any insight as to what might be the best way to go about parallelization tests/making things go faster, please let me know!

**Comment by peterbell10** [comment]
Great to hear this is already being worked on.

> Do you know if `pytest-xdist` has something built in for not running specific tests in parallel or is this just something we'd have to do manually?

I'm not aware of any specific support in `pytest-xdist` but it should be fairly straight forward to do manually. Just have one parallel run with the marked tests deselected, and a second run without parallelism and only the marked tests selected. Similar to how slow tests are handled at the moment.

```bash
# Run parallel tests
OMP_NUM_THREADS=4 pytest test_ops.py -k cpu -m "not noparallel" -n 8
# Run tests marked as noparallel
pytest test_ops.py -k cpu -m noparallel
# Run non-cpu tests
pytest test_ops.py -k "not cpu"
```
